FIX-#5665: Fix operations that flatten an array, as well as handling of where argument in such operations

RehanSD opened the pull request:
<!--
Thank you for your contribution!
Please review the contributing docs: https://modin.readthedocs.io/en/latest/development/contributing.html
if you have questions about contributing.
-->

## What do these changes do?

<!-- Please give a short brief about these changes. -->

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` <!-- you can amend your commit with a signature via `git commit -amend -s` -->
- [x] Resolves #5665 ? <!-- issue must be created for each patch -->
- [x] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date <!-- if you have added, renamed or removed files or directories please update the documentation accordingly -->

anmyachev commented:
@RehanSD this is a fairly large pull request, could you summarize the main changes?

RehanSD commented:
> @RehanSD this is a fairly large pull request, could you summarize the main changes?

Sure! The main changes are as follows:
1. Ensure that all top level Modin.NumPy math functions (eg sum, max, min) can accept Modin.pandas objects (and internally converts them to arrays for operations)
2. Make `__array_ufunc__` handle the `out` and `where` kwargs correctly (it previously ignored them). 
3. Prevent operations on `axis=None` from calling `.flatten()` internally, which creates a bunch of query compilers and is inefficient. 
4. Ensure that operations that take the axis argument handle the `where` kwarg correctly. (For context: on operations like `add` where is a mask that masks the output, so entries where `where` is `False` keep the default value, while for operations like `sum`, `where` is a mask on the input, so entries where `where` is `False` are discarded before summation.)

anmyachev commented on modin/numpy/math.py:
```diff
@@ -44,6 +46,7 @@ def absolute(
 def add(
     x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
 ):
+    x1 = try_convert_from_interoperable_type(x1)
```
There are many functions that are not tested: `add`, `divide`, `float_power`, `floor_divide`, `power`, `prod`, `multiply`, `remainder`, `subtract`, `true_divide`, `mean`, `maximum`, `minimum`, `amax`, `amin`. What is the plan for adding tests?

RehanSD replied:
We test the `array.add`, etc, which these call, so I think that should be ok?

anmyachev replied:
So is this a mistake in `codecov`?

RehanSD replied:
Yeah - I think codecov is complaining that `modin.numpy.add` isn't being tested (which is technically true), but `modin.numpy.add` is literally just a wrapper for `modin.numpy.array.add`, which is tested, so I think its fine not to test the wrapper?

anmyachev replied:
Ok! Thank you for the clarification.

anmyachev commented on modin/numpy/utils.py:
```diff
@@ -0,0 +1,30 @@
+_INTEROPERABLE_TYPES = (pd.DataFrame, pd.Series)
+
+
+def try_convert_from_interoperable_type(obj):
```
It looks like it's not being tested.

RehanSD replied:
I'm not sure we need to test `try_convert` do we? I can add tests if you think its necessary!

anmyachev replied:
At least for one function but `_INTEROPERABLE_TYPES`, yes.

RehanSD replied:
To clarify - you think we should add a test for the `try_convert_from_interoperable_type` function to ensure that it correctly converts convertible objects, and flags for objects that can't be converted? I will add it!

anmyachev replied:
Exactly! Thanks

anmyachev commented on modin/numpy/utils.py:
```diff
@@ -0,0 +1,30 @@
+"""Collection of array utility functions for internal use."""
+
+import modin.pandas as pd
+import modin.numpy as np
```
You are importing a whole `np` module, which itself imports `utils.py` this can very easily spawn cyclic imports.
```suggestion
from modin.numpy.arr import array
```

RehanSD replied:
Makes sense to me!

RehanSD replied:
Tried this and got a circular import error. Reverted!

anmyachev replied:
oops :)

anmyachev commented on modin/numpy/utils.py:
```diff
@@ -0,0 +1,30 @@
+    if isinstance(obj, _INTEROPERABLE_TYPES):
+        new_qc = obj._query_compiler.reset_index(drop=True)
+        new_qc.columns = range(len(new_qc.columns))
+        obj = np.array(
```
```suggestion
        obj = array(
```

anmyachev commented on modin/numpy/math.py:
```diff
@@ -348,12 +362,48 @@ def mean(x1, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
+    x1 = try_convert_from_interoperable_type(x1)
+    if not isinstance(x1, array):
+        ErrorMessage.single_warning(
+            f"Modin NumPy only supports objects of modin.numpy.array types for maximum, not {type(x1)}. Defaulting to NumPy."
```
This error message is very easily parametrized by a function name.

RehanSD replied:
Sounds good to me!

RehanSD replied:
Added a method to `ErrorMessage` class to refactor this!

anmyachev commented on modin/numpy/math.py:
```diff
@@ -348,12 +362,48 @@ def mean(x1, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
+def minimum(
+    x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
+):
+    x1 = try_convert_from_interoperable_type(x1)
```
```suggestion
    x1 = try_convert_from_interoperable_type(x1)
    x2 = try_convert_from_interoperable_type(x2)
```

RehanSD replied:
We don't need to convert the second operand, since we do so inside the array class!

anmyachev replied:
Apparently I do not fully understand how the code works.

Could you explain where we convert the second operand?
And why can't we convert the first operand there?

At least the next check can be rewritten as follows:
```python
if not isinstance(x1, (array, pd.DataFrame, pd.Series)):
```

RehanSD replied:
So all of the `modin.numpy` methods are simply wrappers that internally call `modin.numpy.array.<method>`. E.g.:
```python
def add(...):
x1 = try_convert_from_interoperable_type(x1)
if isinstance(x1, array):
x1.add ...
```
So we need to make sure that the first operand is converted from a modin pandas object to a modin numpy object so that it can call the appropriate method, but inside the `modin.numpy.array.add` method, the first line attempts to convert the other operand to an array.

Take `__add__` as an example. On [this](https://github.com/RehanSD/modin/blob/a94b84cd6ebc63a0d951476ee1fd34629ae7c6de/modin/numpy/arr.py#L868) line, we call `_binary_op` - `_binary_op`, makes sure that the second operand is of type array, and errors if it isn't on these lines: https://github.com/RehanSD/modin/blob/a94b84cd6ebc63a0d951476ee1fd34629ae7c6de/modin/numpy/arr.py#L634-L638 - so we don't need to convert the second operand at the numpy layer.

anmyachev replied:
Got it! Thank you for the clarification.

anmyachev commented on modin/numpy/math.py:
```diff
@@ -348,12 +362,48 @@ def mean(x1, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
+def maximum(
+    x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
+):
+    x1 = try_convert_from_interoperable_type(x1)
```
```suggestion
    x1 = try_convert_from_interoperable_type(x1)
    x2 = try_convert_from_interoperable_type(x2)
```

anmyachev commented on modin/numpy/math.py:
```diff
@@ -309,6 +321,7 @@ def sum(arr, axis=None, dtype=None, out=None, keepdims=None, initial=None, where
 def true_divide(
     x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
 ):
+    x1 = try_convert_from_interoperable_type(x1)
```
```suggestion
    x1 = try_convert_from_interoperable_type(x1)
    x2 = try_convert_from_interoperable_type(x2)
```

anmyachev commented on modin/numpy/math.py:
```diff
@@ -262,6 +272,7 @@ def remainder(
 def subtract(
     x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
 ):
+    x1 = try_convert_from_interoperable_type(x1)
```
```suggestion
    x1 = try_convert_from_interoperable_type(x1)
    x2 = try_convert_from_interoperable_type(x2)
```

anmyachev commented on modin/numpy/math.py:
```diff
@@ -231,6 +240,7 @@ def multiply(
 def remainder(
     x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
 ):
+    x1 = try_convert_from_interoperable_type(x1)
```
```suggestion
    x1 = try_convert_from_interoperable_type(x1)
    x2 = try_convert_from_interoperable_type(x2)
```

anmyachev commented on modin/numpy/math.py:
```diff
@@ -203,6 +211,7 @@ def prod(a, axis=None, out=None, keepdims=None, where=True, dtype=None, initial=
 def multiply(
     x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
 ):
+    x1 = try_convert_from_interoperable_type(x1)
```
```suggestion
    x1 = try_convert_from_interoperable_type(x1)
    x2 = try_convert_from_interoperable_type(x2)
```

anmyachev commented on modin/numpy/math.py:
```diff
@@ -156,6 +162,7 @@ def floor_divide(
 def power(
     x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
 ):
+    x1 = try_convert_from_interoperable_type(x1)
```
```suggestion
    x1 = try_convert_from_interoperable_type(x1)
    x2 = try_convert_from_interoperable_type(x2)
```

anmyachev commented on modin/numpy/math.py:
```diff
@@ -128,6 +133,7 @@ def float_power(
 def floor_divide(
     x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
 ):
+    x1 = try_convert_from_interoperable_type(x1)
```
```suggestion
    x1 = try_convert_from_interoperable_type(x1)
    x2 = try_convert_from_interoperable_type(x2)
```

anmyachev commented on modin/numpy/math.py:
```diff
@@ -100,6 +104,7 @@ def divide(
 def float_power(
     x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
 ):
+    x1 = try_convert_from_interoperable_type(x1)
```
```suggestion
    x1 = try_convert_from_interoperable_type(x1)
    x2 = try_convert_from_interoperable_type(x2)
```

anmyachev commented on modin/numpy/math.py:
```diff
@@ -72,6 +75,7 @@ def add(
 def divide(
     x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
 ):
+    x1 = try_convert_from_interoperable_type(x1)
```
```suggestion
    x1 = try_convert_from_interoperable_type(x1)
    x2 = try_convert_from_interoperable_type(x2)
```

anmyachev commented on modin/numpy/math.py:
```diff
@@ -44,6 +46,7 @@ def absolute(
 def add(
     x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
 ):
+    x1 = try_convert_from_interoperable_type(x1)
```
```suggestion
    x1 = try_convert_from_interoperable_type(x1)
    x2 = try_convert_from_interoperable_type(x2)
```

anmyachev commented on modin/numpy/arr.py:
```diff
@@ -294,7 +282,20 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
-        return array(_query_compiler=new_ufunc(*args, **kwargs), _ndim=out_ndim)
+        out_kwarg = kwargs.get("out", None)
+        if out_kwarg is not None:
+            out_kwarg = out_kwarg[0]
```
Why?

RehanSD replied:
If the `out_kwarg` is a modin.numpy.array, getting it from kwargs returns a 1-length tuple containing the array. I'll add a comment clarifying this as well!

anmyachev commented on modin/numpy/arr.py:
```diff
@@ -408,71 +409,88 @@ def max(
+            if not isinstance(where, array):
+                result = self._query_compiler.max(axis=0)
+            else:
+                result = where.where(self, initial)._query_compiler.max(axis=0)
```
I think this is an important case, we need to test it.

RehanSD replied:
Makes sense to me!

anmyachev commented on modin/numpy/arr.py:
```diff
@@ -704,50 +741,101 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
         if out is not None:
-            out._query_compiler = (out * numpy.nan).astype(out_dtype)._query_compiler
-        return fix_dtypes_and_determine_return(result, new_ndim, dtype, out, where)
+            out._query_compiler = array(
+                numpy.ones(out.shape) * numpy.nan
+            )._query_compiler
```
Why is the `array` call here?
What is the difference with `out._query_compiler = (numpy.ones_like(out) * initial)._query_compiler`?

anmyachev replied:
Also why is `.astype(out_dtype)` not called now?

RehanSD replied:
The main difference is in the dtype of the created array. I changed it to this because I thought that the old dtype was causing the ubuntu test to fail, but it was actually the as type part - which was causing it to fail since we had an array of NaN's and tried to convert that to an array of dtype int, which was failing.

RehanSD commented:
Thank you for all of the reviews @anmyachev!

anmyachev commented on modin/numpy/arr.py:
```diff
@@ -704,50 +741,101 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
         if out is not None:
-            out._query_compiler = (out * numpy.nan).astype(out_dtype)._query_compiler
-        return fix_dtypes_and_determine_return(result, new_ndim, dtype, out, where)
+            out._query_compiler = array(
+                numpy.ones(out.shape) * numpy.nan
+            )._query_compiler
```
Will it work? More consistent with other places, of course, if it does not bring any undesirable effects in this case or slow it down.
```suggestion
            out._query_compiler = (numpy.ones_like(out) * numpy.nan)._query_compiler
```

RehanSD replied:
I don't think it will slow anything down - I'll go ahead and change it - good catch, thank you!

anmyachev commented on modin/numpy/arr.py:
```diff
@@ -704,50 +741,101 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
                     )
-                return fix_dtypes_and_determine_return(result, 1, dtype, out, where)
+                if out is not None:
+                    out._query_compiler = array(
+                        numpy.ones(out.shape) * numpy.nan
+                    )._query_compiler
```
Will it work?
```suggestion
                    out._query_compiler = (numpy.ones_like(out) * numpy.nan)._query_compiler
```

anmyachev commented on modin/numpy/arr.py:
```diff
@@ -704,50 +741,101 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
+                        f"operand was set up as a reduction along axis 1, but the length of the axis is {out.shape[0]} (it has to be 1)"
+                    )
+                if out is not None:
+                    out._query_compiler = array(
+                        numpy.ones(out.shape) * numpy.nan
+                    )._query_compiler
```
Will it work?
```suggestion
                    out._query_compiler = (numpy.ones_like(out) * numpy.nan)._query_compiler
```

anmyachev commented on modin/numpy/arr.py:
```diff
@@ -408,71 +411,88 @@ def max(
+            result = self._query_compiler
+            if isinstance(where, array):
+                result = where.where(self, initial)._query_compiler
+            result = result.max(axis=0).max(axis=1).to_numpy().flatten()[0]
```
Will it work?
```suggestion
            result = result.max().to_numpy().flatten()[0]
```

RehanSD replied:
This won't work - the query compiler expects an axis argument to functions like `max` and `min`. If none is provided, it errors out. (this is because `df.max(axis=None)` is the same as `df.max(axis=0)`, unlike with NumPy, where `axis=None` means a reduction to a single value).

anmyachev commented on modin/numpy/arr.py:
```diff
@@ -485,71 +505,88 @@ def min(
+            result = self._query_compiler
+            if isinstance(where, array):
+                result = where.where(self, initial)._query_compiler
+            result = result.min(axis=0).min(axis=1).to_numpy().flatten()[0]
```
Will it work?
```suggestion
            result = result.min().to_numpy().flatten()[0]
```

anmyachev commented on modin/numpy/arr.py:
```diff
@@ -408,71 +411,88 @@ def max(
             if axis == 1:
                 raise numpy.AxisError(1, 1)
-            result = self._query_compiler.max(axis=0)
+            if not isinstance(where, array):
+                result = self._query_compiler.max(axis=0)
+            else:
+                result = where.where(self, initial)._query_compiler.max(axis=0)
```
```suggestion
            if isinstance(where, array):
                result = where.where(self, initial)._query_compiler
            result = result.max(axis=0)
```

anmyachev commented on modin/numpy/arr.py:
```diff
@@ -485,71 +505,88 @@ def min(
             if axis == 1:
                 raise numpy.AxisError(1, 1)
-            result = self._query_compiler.min(axis=0)
+            if not isinstance(where, array):
+                result = self._query_compiler.min(axis=0)
+            else:
+                result = where.where(self, initial)._query_compiler.min(axis=0)
```
```suggestion
            if isinstance(where, array):
                result = where.where(self, initial)._query_compiler
            result = result.min(axis=0)
```

anmyachev commented:
@RehanSD these are some of my last comments, please answer them and if no one minds, I'll merge it on Monday evening.

RehanSD commented:
Thank you @anmyachev - really appreciate the help getting this over the finish line! I've gone ahead and addressed your comments, and am pushing a revised version of the code!

vnlitvinov commented:
> and if no one minds, I'll merge it on Monday evening.

Do note that as of right now, CI is minding the merge 😉

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -294,7 +282,22 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
-        return array(_query_compiler=new_ufunc(*args, **kwargs), _ndim=out_ndim)
+        out_kwarg = kwargs.get("out", None)
+        if out_kwarg is not None:
+            # If `out` is a modin.numpy.array, `kwargs.get("out")` returns a 1-tuple
```
but what if `out` is _not_ a `modin.numpy.array`?.. shouldn't we `isinstance()` check it instead of `is not None`?

RehanSD replied:
That's handled by the function `fix_dtypes_and_determine_return`. It calls `check_can_broadcast_to_output`, which first checks if the out variable is a modin.numpy.array, and errors if it is not!

vnlitvinov replied:
TBH I still don't understand why do we pass a variable using a 1-element tuple... feels super-weird. Could you pls explain?

RehanSD replied:
Sorry - to clarify, we don't do that - that's just how NumPy handles the arguments. Say someone calls `numpy.tanh(modin.numpy.array, *args, **kwargs)` - the following happens:
1. NumPy recognizes that our `modin.numpy.array` object has an `__array_ufunc__` method.
2. NumPy translates `numpy.tanh` to `modin.numpy.array.__array_ufunc__(numpy.tanh, "__call__", *args, **kwargs)`
Our `out` argument is inside the `**kwargs`, but the packaging (it being inside a 1-tuple) is handled by NumPy, so I just unpack it if its 1-tuple.

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -408,71 +411,88 @@ def max(
             if axis == 1:
                 raise numpy.AxisError(1, 1)
-            result = self._query_compiler.max(axis=0)
+            result = self._query_compiler
+            if isinstance(where, array):
+                result = where.where(self, initial)._query_compiler
+            result = result.max(axis=0)
```
```suggestion
            target = where.where(self, initial) if isinstance(where, array) else self
            result = target._query_compiler.max(axis=0)
```
?

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -408,71 +411,88 @@ def max(
                         f"operand was set up as a reduction along axis 0, but the length of the axis is {out.shape[0]} (it has to be 1)"
                     )
-                return fix_dtypes_and_determine_return(result, 1, dtype, out, where)
+                return (
+                    fix_dtypes_and_determine_return(
+                        result, 1, dtype, out, (where is not False)
+                    )
+                    if where is not False or out is not None
+                    else array([initial])
+                )
```
```suggestion
                if where is not False or out is not None:
                    return fix_dtypes_and_determine_return(result, 1, dtype, out, where is not False)
                return array([initial])
```
I wonder if this reads better...

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -408,71 +411,88 @@ def max(
-                initial=initial,
-                where=where,
-            )
+            result = self._query_compiler
+            if isinstance(where, array):
+                result = where.where(self, initial)._query_compiler
```
```suggestion
            result = (where.where(self, initial) if isinstance(where, array) else self)._query_compiler
```
?
or like in my suggestion for `max()` with creation of intermediate `target`

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -408,71 +411,88 @@ def max(
                 if out is not None and out.shape != (1, 1):
                     raise ValueError(
-                        f"operand was set up as a reduction along axis 0, but the length of the axis is {out.shape[0]} (it has to be 1)"
+                        f"operand was set up as a reduction along axis 1, but the length of the axis is {out.shape[0]} (it has to be 1)"
```
this error looks misleading, it talks about axis 1, reports axis 0, and _really checks_ both axes 😆

RehanSD replied:
True, but in this case I was just copying the error NumPy gives.

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -408,71 +411,88 @@ def max(
                     )
-                return fix_dtypes_and_determine_return(
-                    array(numpy.array([[result]]))._query_compiler, 2, dtype, out, where
+                    if where is not False or out is not None
```
I'm not sure I like this `a if foo else b` inside the `return` statement when it's so lengthy... I personally like it done like 
```python
if foo:
    return a
return b
```

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -408,71 +411,88 @@ def max(
-            return result
-        result = self._query_compiler.max(axis=axis)
+            return result if where is not False else initial
+        result = self._query_compiler
+        if isinstance(where, array):
+            result = where.where(self, initial)._query_compiler
+        result = result.max(axis=axis)
```
```suggestion
        target = where.where(self, initial) if isinstance(where, array) else self
        result = target._query_compiler.max(axis=axis)
```

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -408,71 +411,88 @@ def max(
+            out._query_compiler = (numpy.ones_like(out) * initial)._query_compiler
         intermediate = fix_dtypes_and_determine_return(
-            result, new_ndim, dtype, out, where
+            result, new_ndim, dtype, out, (where is not False)
```
why do we need braces around the `where is not False`?

RehanSD replied:
I added them for clarity - it's not really necessary, just a personal style thing.

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -485,71 +505,88 @@ def min(
         if self._ndim == 1:
             if axis == 1:
                 raise numpy.AxisError(1, 1)
-            result = self._query_compiler.min(axis=0)
```
starting from here, it looks like a copy of `.max()` but with `min()` instead... why did we copy-paste so much?

I won't repeat my suggestions on `target = where.where(...) if foo else self` here, but they do apply

RehanSD replied:
A lot of the code is very similar with some small changes - I think @noloerino is planning a refactor for the future!

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -485,71 +505,88 @@ def min(
+            result = result.min(axis=0)
             if keepdims:
-                if initial is not None and result.lt(initial):
+                if initial is not None and result.lt(initial).any():
```
```suggestion
                if initial is not None and result.gt(initial).any():
```
I wonder why we `.lt()` both in `.max()` and in `.min()`, feels like it should be `.gt()` here?..

RehanSD replied:
Good catch, thank you!

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -704,50 +741,99 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
             if dtype is not None
             else (out.dtype if out is not None else self.dtype)
         )
+        out_type = out_dtype if not hasattr(out_dtype, "type") else out_dtype.type
```
```suggestion
        out_type = getattr(out_dtype, "type", out_dtype)
```
😉 
does the same, works slightly faster (because `hasattr()` internally does the `getattr()` thing and returns True if no error happened)

RehanSD replied:
Fixed!

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -704,50 +741,99 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
             else (out.dtype if out is not None else self.dtype)
         )
+        out_type = out_dtype if not hasattr(out_dtype, "type") else out_dtype.type
+        if isinstance(where, array) and issubclass(out_type, numpy.integer):
+            out_dtype = numpy.float64
```
why?..

RehanSD replied:
if we try and convert out to the out dtype on ubuntu, it errors, since out will contain NaNs.

vnlitvinov replied:
As far as I can see, this handles the case when after `where.where()` we have some `NaN`-s inside... but it's not always going to happen? What does numpy do in this case, does it convert the output to `float` regardless of whether `NaN`-s are present?

P.S. I don't understand what "ubuntu" means here 🙃 so I "played by ear" and guessed what you meant, hopefully correctly...

RehanSD replied:
Sorry - to clarify, I meant that if we don't do this, our tests fail on the Ubuntu OS (but not on macOS or windows).

Interesting - I just checked `NumPy`, and it looks like if `NaN` is present in the array, it returns `NaN` as the mean (or a junk value if the output dtype is int). We ignore the `NaN` and compute the mean or sum as if it wasn't there - because that's how pandas does it, but NumPy propagates the NaN.

Do you think we should try and resolve this in this PR? I'm concerned its already kind of big, and if we do decide to fix this, this will also mean we need to handle our masks differently, since we currently take advantage of pandas ignoring `NaN` behavior to replace values we're making with `NaN` without affecting the sum, so we'd probably need to come up with a new method for masking. (Although we could technically convert this to a Binary algebra operator maybe? Where the second operand is the where array for masking?)

vnlitvinov replied:
I would say, let's make a separate issue, as this PR is really already big enough 🙃

RehanSD replied:
Issue created: https://github.com/modin-project/modin/issues/5720

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -704,50 +741,99 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
                 raise numpy.AxisError(1, 1)
-            result = self._query_compiler.astype(
-                {col_name: out_dtype for col_name in self._query_compiler.columns}
+            result = self._query_compiler
+            if isinstance(where, array):
+                result = where.where(self, numpy.nan)._query_compiler
+            result = result.astype(
```
```suggestion
            target = where.where(self, numpy.nan) if isinstance(where, array) else self
            result = target._query_compiler.astype(
```

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -704,50 +741,99 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
+                    if where is not False or out is not None
+                    else array([numpy.nan], dtype=out_dtype)
+                )
             return result.to_numpy()[0, 0] if where else numpy.nan
```
shouldn't we change this to `if where is not False` as well?

RehanSD replied:
No - in this case, where should just be evaluated for truthiness, since we're returning a single value.

vnlitvinov replied:
I wonder if this should be a comment in the code, as this is the only place where the pattern is different...

RehanSD replied:
Sure - I can add a comment!

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -704,50 +741,99 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
+                    [result._query_compiler.sum(axis=1).sum(axis=0).to_numpy()[0, 0]],
+                    dtype=out_dtype,
+                )
+                / (
+                    prod(self.shape)
+                    - result._query_compiler.isna()
+                    .sum(axis=1)
+                    .sum(axis=0)
+                    .to_numpy()[0, 0]
+                )
```
I wonder if extracting the calculation of "now many NAs is there" to a separate variable would improve the readability... my first impression was that this piece was accidentally calling `.sum()` on only part of the equation, until I realized what was going on

RehanSD replied:
I can break this up and add some comments to help explain what's going on!

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -704,50 +741,99 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
-            return result
-        result = self._query_compiler.astype(
+            return result if where is not False else numpy.nan
+        result = self._query_compiler
+        if isinstance(where, array):
+            result = where.where(self, numpy.nan)._query_compiler
+        result = result.astype(
```
```suggestion
        target = where.where(self, numpy.nan) if isinstance(where, array) else self
        result = target._query_compiler.astype(
```

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -1020,74 +1106,94 @@ def prod(
                 raise numpy.AxisError(1, 1)
-            result = self._query_compiler.astype(
-                {col_name: out_dtype for col_name in self._query_compiler.columns}
+            result = self._query_compiler
+            if isinstance(where, array):
+                result = where.where(self, 1)._query_compiler
+            result = result.astype(
```
I will stop my suggestion of `target = where.where(...) if foo else self` from now on, but it does apply here and in other places

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -1020,74 +1106,94 @@ def prod(
         if self._ndim == 1:
             if axis == 1:
                 raise numpy.AxisError(1, 1)
-            result = self._query_compiler.astype(
-                {col_name: out_dtype for col_name in self._query_compiler.columns}
+            result = self._query_compiler
+            if isinstance(where, array):
+                result = where.where(self, 1)._query_compiler
+            result = result.astype(
+                {col_name: out_dtype for col_name in result.columns}
             ).prod(axis=0)
-            if initial is not None:
-                result = result.mul(initial)
+            result = result.mul(initial)
```
what would happen if `initial` is None?

RehanSD replied:
I think it would fail, but we set `initial` to 1 if it is `None` at the top, so it doesn't matter.

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -1310,72 +1416,93 @@ def sum(
         if self._ndim == 1:
             if axis == 1:
                 raise numpy.AxisError(1, 1)
-            result = self._query_compiler.astype(
-                {col_name: out_dtype for col_name in self._query_compiler.columns}
+            result = self._query_compiler
+            if isinstance(where, array):
+                result = where.where(self, 0)._query_compiler
+            result = result.astype(
+                {col_name: out_dtype for col_name in result.columns}
             ).sum(axis=0)
-            if initial is not None:
-                result = result.add(initial)
+            result = result.add(initial)
```
what would happen if `initial` is None?

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -1310,72 +1416,93 @@ def sum(
+                ._query_compiler.sum(axis=1)
+                .sum(axis=0)
+                .to_numpy()
+                .flatten()[0]
```
why do we need to `.flatten()`, cannot we just `[0, 0]` here? it should be reduced to a single cell anyway

vnlitvinov commented on modin/numpy/math.py:
```diff
@@ -44,10 +44,9 @@ def absolute(
-        ErrorMessage.single_warning(
-            f"Modin NumPy only supports objects of modin.numpy.array types for add, not {type(x1)}. Defaulting to NumPy."
-        )
+        ErrorMessage.bad_type_for_numpy_op("add", type(x1))
```
why don't we convert `x2` the same way? here and for other binary functions

RehanSD replied:
We don't need to - since the `binary_op` code takes care of converting x2!

vnlitvinov commented on modin/numpy/math.py:
```diff
@@ -14,16 +14,16 @@
 def absolute(
     x, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
 ):
+    x = try_convert_from_interoperable_type(x)
     if not isinstance(x, array):
-        ErrorMessage.single_warning(
-            f"Modin NumPy only supports objects of modin.numpy.array types for absolute, not {type(x)}. Defaulting to NumPy."
-        )
+        ErrorMessage.bad_type_for_numpy_op("absolute", type(x))
```
I wonder if this should be hidden in a decorator, like this (untested):
```python
from functools import wraps

def _default_to_numpy(func):
    @wraps(func):
    def wrapped(arr, *args, **kw):
        arr = try_convert_from_interoperable_type(arr)
        if not isinstance(arr, array):
            ErrorMessage.bad_type_for_numpy_op(func.__name__, type(arr))
            return getattr(numpy, func.__name__)(arr, *args, **kw)
        return func(arr, *args, **kw)
    return wrapped


@_default_to_numpy
def absolute(
    x, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
):
    return x.absolute(
        out=out, where=where, casting=casting, order=order, dtype=dtype, subok=subok
    )

@_default_to_numpy
def add(
    x1, x2, out=None, where=True, casting="same_kind", order="K", dtype=None, subok=True
):
    return x1.__add__(
        x2,
        out=out,
        where=where,
        casting=casting,
        order=order,
        dtype=dtype,
        subok=subok,
    )

# etc.
```
?

RehanSD replied:
Perhaps - I think its little enough code that its fine for now IMO!

vnlitvinov replied:
I guess this can be handled during the refactor you and @noloerino were planning ahead...

vnlitvinov commented on modin/numpy/math.py:
```diff
@@ -348,16 +336,46 @@ def mean(x1, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
 # Maximum and minimum are ufunc's in NumPy, which means that our array's __array_ufunc__
 # implementation will automatically handle this, so we can just use NumPy's maximum/minimum
 # since that will route to our array's ufunc.
```
this comment seems outdated now

RehanSD replied:
Fixed!

vnlitvinov commented on modin/numpy/test/test_array_axis_functions.py:
```diff
@@ -122,6 +137,13 @@ def test_min():
+    modin_arr = np.array(numpy_arr)
+    numpy_mask = numpy.array([True, False, True, True, True, True])
+    modin_mask = np.array(numpy_mask)
+    assert numpy_arr.max(where=numpy_mask, initial=5) == modin_arr.max(
```
```suggestion
    assert numpy_arr.min(where=numpy_mask, initial=5) == modin_arr.min(
```
I don't see a point of testing .`max()` within `test_min()` function 😉

RehanSD replied:
Good catch, thank you!

vnlitvinov commented on modin/numpy/test/test_array_axis_functions.py:
```diff
@@ -189,6 +211,14 @@ def test_min():
+    numpy_mask = numpy.array([[True, False, True], [True, True, True]])
+    modin_mask = np.array(numpy_mask)
+    numpy.testing.assert_equal(
+        numpy_arr.max(where=numpy_mask, initial=5),
+        modin_arr.max(where=modin_mask, initial=5),
```
```suggestion
        numpy_arr.min(where=numpy_mask, initial=5),
        modin_arr.min(where=modin_mask, initial=5),
```

RehanSD replied:
Good catch, thank you!

vnlitvinov commented:
I should've posted a summary... most of my suggestions are a matter of taste, but I believe that `.min()` implementation is bugged when `initial` is not None (https://github.com/modin-project/modin/pull/5668#discussion_r1118326375), and I'm sure tests for `.min()` have a copy-paste error as they're testing for `.max()` instead

RehanSD commented:
Thank you for the review @vnlitvinov! I've gone ahead and addressed your comments!

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -426,23 +424,20 @@ def max(
-                    if where is not False or out is not None
-                    else array([initial])
-                )
+                else:
+                    return array([initial])
```
technically you can do this:
```python
if where is not False or out is not None:
    return fix_dtypes_and_determine_return(...)
return array([initial])
```
i.e. get rid of `else:` line completely... though I think that it may actually read clearer your way now

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -763,31 +744,33 @@ def mean(self, axis=None, dtype=None, out=None, keepdims=None, *, where=True):
-                )
-            ).flatten()[0]
+                / num_total_elements
+            )[0]
```
now I wonder what is the reason of making a single-value numpy array to only take its sole element straight away... cannot we do it like
```python
result = result._query_compiler.sum(axis=1).sum(axis=0).to_numpy()[0, 0] / num_total_elements
```
?

RehanSD replied:
I wanted to keep it a NumPy array so that we can get NumPy division rules - say we have 0 non-na elements in the list - rather than having an error (because we're dividing by 0), NumPy will handle the division correctly, and give us the appropriate result (i.e. NaN vs. INF depending on what our output dtype should be)

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -1106,9 +1085,8 @@ def prod(
-                result = where.where(self, 1)._query_compiler
+            target = where.where(self, 1) if isinstance(where, array) else self
+            result = target._query_compiler
             result = result.astype(
```
```python
result = target._query_compiler.astype(
```
?

RehanSD replied:
Fixed!

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -1165,13 +1141,11 @@ def prod(
-            result = where.where(self, 1)._query_compiler
+        target = where.where(self, 1) if isinstance(where, array) else self
+        result = target._query_compiler
         result = result.astype(
```
```python
result = target._query_compiler.astype(
```
?

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -1416,9 +1389,8 @@ def sum(
-                result = where.where(self, 0)._query_compiler
+            target = where.where(self, 0) if isinstance(where, array) else self
+            result = target._query_compiler
             result = result.astype(
```
```python
result = target._query_compiler.astype(
```
?

vnlitvinov commented on modin/numpy/arr.py:
```diff
@@ -1465,21 +1435,19 @@ def sum(
-            result = where.where(self, 0)._query_compiler
+        target = where.where(self, 0) if isinstance(where, array) else self
+        result = target._query_compiler
         result = result.astype(
```
```python
result = target._query_compiler.astype(
```
?

vnlitvinov reviewed (approved):
Taste changes and questions left... bugs have been squashed, so I'm stamping a green mark. 
🚢!

vnlitvinov reviewed (approved):
Awesome, LGTM!
